File: 3903-smallest-stable-index-i/3903-smallest-stable-index-i.cpp
class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
        int res = nums.size();
        vector<int> pre(nums.size(), 0);
        int minNum = INT_MAX;
        int curMax = 0;
        int curMin = INT_MAX;
        int r = nums.size() - 1;
        for (int l = 0; l < nums.size(); l++) {
            curMax = max(curMax, nums[l]);
            curMin = min(curMin, nums[r]);
            pre[l] += curMax;
            pre[r] -= curMin;
            if (l >= r) {
                if (pre[l] <= k) {
                    res = min(res, l);
                }
                if (pre[r] <= k) {
                    res = min(res, r);
                }
            }
            r--;
        }
        return res != nums.size() ? res : -1;
    }
};

File: 3903-smallest-stable-index-i/3903-smallest-stable-index-i.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        res = len(nums)
        pre = [0 for _ in range(len(nums))]
        minNum = float('inf')
        curMax = 0
        curMin = float('inf')
        r = len(nums) - 1
        for l in range(len(nums)):
            curMax = max(curMax, nums[l])
            curMin = min(curMin, nums[r])
            pre[l] += curMax
            pre[r] -= curMin
            if l >= r:
                if pre[l] <= k:
                    res = min(res, l)
                if pre[r] <= k:
                    res = min(res, r)
            r -= 1
        return res if res != len(nums) else -1